internal/precache/precache.go

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2023 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this inputFilePath except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package precache
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// +kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch;create;update;delete
+// +kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;watch;create;update;delete
+
+var log = ctrl.Log.WithName("precache")
+
+// Config defines the configuration options for a pre-caching job.
+type Config struct {
+	ImageList          []string
+	NumConcurrentPulls int
+	// To run pre-caching job with an adjusted niceness, which affects process scheduling.
+	// Niceness values range from -20 (most favorable to the process) to 19 (least favorable to the process).
+	NicePriority int
+	// To configure the I/O-scheduling class and priority of a process.
+	IoNiceClass    int // 0: none, 1: realtime, 2: best-effort, 3: idle
+	IoNicePriority int // priority (0..7) in the specified scheduling class, only for the realtime and best-effort classes
+}
+
+// NewConfig creates a new Config instance with the provided imageList and optional configuration parameters.
+// It initializes the Config with default values and updates specific fields using key-value pairs in args.
+// Supported configuration options in args:
+//   - "NumConcurrentPulls" (int): Number of concurrent pulls for pre-caching.
+//   - "NicePriority" (int): Nice priority for pre-caching.
+//   - "IoNiceClass" (int): I/O nice class for pre-caching.
+//   - "IoNicePriority" (int): I/O nice priority for pre-caching.
+//
+// Example usage:
+//
+//	config := NewConfig(imageList, "NumConcurrentPulls", 10, "NicePriority", 5)
+func NewConfig(imageList []string, args ...interface{}) *Config {
```

Comment:
Curious why you're allowing arbitrary args and not something like a struct which would just be type checked etc? 

Generally the pattern i have noticed `func NewConfig(options *PrecacheOptions) *PrecacheConfig`

Reply:
Implemented this as a factory function for creating a new Config instance by allowing the user to specify a subset of arguments without having to specify the entire parameter list. This then allows for default values to be assigned for non-specified parameters.